pkg/master/reconcilers/lease.go

```diff
@@ -69,7 +69,9 @@ func (s *storageLeases) ListLeases() ([]string, error) {
 
 	ipList := make([]string, len(ipInfoList.Items))
 	for i, ip := range ipInfoList.Items {
-		ipList[i] = ip.Subsets[0].Addresses[0].IP
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 {
```

Comment:
since this is now conditional, make `ipList` length 0 and append

Reply:
and is there a test that covers this?

Reply:
Since ip.Subsets is non-empty is the majority of cases, I think leaving the length of ipList as what it was would be better.

Reply:
returning a string array containing empty strings seems problematic. you can preallocate the desired capacity, but make the length 0 to avoid empty strings.

Reply:
Trying to construct a test where storageLeases#ListLeases is called.
Without populating storage field, I got:
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x1dff413]

goroutine 25 [running]:
testing.tRunner.func1(0xc00037e300)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0x1f19f20, 0x2f2a5e0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
k8s.io/kubernetes/pkg/master/reconcilers.(*storageLeases).ListLeases(0xc00007af70, 0x100000001, 0xc00006b770, 0x104ef38, 0x2c590e1a5cc8, 0xc645148)
	/Users/yute/go-workspace/src/k8s.io/kubernetes/pkg/master/reconcilers/lease.go:66 +0x63
```
One option is duplicating code from Config#createLeaseReconciler but I am not sure that's the way to go.

Reply:
StorageFactory is an interface:
```
type StorageFactory interface {
```
Trying to figure out how to get StorageFactory instance.

Reply:
Attempted the following change to pkg/master/master_test.go
```
	c, err := config.ExtraConfig.StorageFactory.NewConfig(api.Resource("apiServerIPInfo"))
	if err != nil {
		t.Errorf("Error determining service IP ranges: %v", err)
	}
	leaseStorage, _, err := storagefactory.Create(*c)
	if err != nil {
		t.Errorf("Error creating storage factory: %v", err)
	}
	masterLeases := reconcilers.NewLeases(leaseStorage, "/masterleases/", time.Minute)
	_, err = masterLeases.ListLeases()
	if err != nil {
		t.Errorf("encountered %v", err)
	}
```

Reply:
Making the length 0 resulted in (services.log):
```
E1107 16:24:46.044672    1958 runtime.go:78] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 3118 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x4a435e0, 0xc00076a240)
...
        k8s.io/kubernetes/pkg/master/reconcilers.(*storageLeases).ListLeases(0xc0010d6120, 0xc00a3e6b40, 0xc011ea6640, 0x8, 0xa, 0x0)     /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/reconcilers/lease.go:73 +0x311
```